Mathlib/Algebra/Homology/BifunctorHomotopy.lean

```diff
@@ -0,0 +1,162 @@
+    rw [← hj, ← ComplexShape.next_π₂ c₁ c i₁ hi₂, ComplexShape.next_π₁ c₂ c hi₁ i₂]
+  rw [HomologicalComplex₂.d₂_eq _ _ _ hi₂ _ hj', HomologicalComplex₂.d₂_eq _ _ _ hi₂ _
+        (by rw [← c.next_eq' (ComplexShape.rel_π₂ c₁ c i₁' hi₂), hj]),
+    Linear.comp_units_smul, Linear.comp_units_smul, Linear.units_smul_comp, assoc,
+    ιMapBifunctor_hom₁ _ _ _ _ _ _ _ _ _ _ (c₁.prev_eq' hi₁),
+    ιMapBifunctorOrZero_eq _ _ _ _ _ _ _ hj',
+    Linear.comp_units_smul, smul_smul, smul_smul]
+  dsimp
+  rw [NatTrans.naturality_assoc, ComplexShape.ε₁_ε₂ c hi₁ hi₂, neg_mul, Units.neg_smul, neg_inj,
+    smul_left_cancel_iff, ← Functor.map_comp_assoc, ← Functor.map_comp_assoc, f₂.comm]
+  rfl
+
+lemma comm₁ (j : J) :
+    (mapBifunctorMap f₁ f₂ F c).f j =
+    (mapBifunctor K₁ K₂ F c).d j (c.next j) ≫
+          mapBifunctorMapHomotopy.hom₁ h₁ f₂ F c (c.next j) j +
+        mapBifunctorMapHomotopy.hom₁ h₁ f₂ F c j (c.prev j) ≫
+          (mapBifunctor L₁ L₂ F c).d (c.prev j) j +
+      (mapBifunctorMap f₁' f₂ F c).f j := by
+  ext i₁ i₂ h
+  dsimp [ιMapBifunctor, mapBifunctor, mapBifunctorMap]
+  simp? [h₁.comm i₁, dFrom, fromNext, toPrev, dTo] says
+    simp only [HomologicalComplex₂.ιTotal_map, Functor.mapBifunctorHomologicalComplex_obj_obj_X_X,
+    HomologicalComplex₂.total_X, Functor.mapBifunctorHomologicalComplex_obj_obj_toGradedObject,
+    comp_f, Functor.mapBifunctorHomologicalComplex_map_app_f_f, h₁.comm i₁,
+    dNext_eq_dFrom_fromNext, dFrom, fromNext, AddMonoidHom.mk'_apply, prevD_eq_toPrev_dTo,
+    toPrev, dTo, Functor.map_add, Functor.map_comp, NatTrans.app_add, NatTrans.comp_app,
+    Functor.mapBifunctorHomologicalComplex_obj_map_f_f, Preadditive.add_comp, assoc,
+    Preadditive.comp_add, HomologicalComplex₂.ι_D₁_assoc, GradedObject.mapBifunctor_obj_obj,
+    HomologicalComplex₂.ι_D₂_assoc, add_left_inj]
+  have : ∀ {X Y : D} (a b c d e f : X ⟶ Y), a = c → b = e → f = -d →
+      a + b = c + d + (e + f) := by rintro X Y a b _ d _ _ rfl rfl rfl; abel
+  apply this
+  · by_cases h₃ : c₁.Rel i₁ (c₁.next i₁)
+    · rw [HomologicalComplex₂.d₁_eq _ _ h₃ _ _ (by rw [← h, ComplexShape.next_π₁ c₂ c h₃])]
+      dsimp
+      rw [Linear.units_smul_comp, assoc]
+      erw [ιMapBifunctor_hom₁ _ _ _ _ i₁ _ _ _ _ _ (c₁.prev_eq' h₃)]
+      rw [Linear.comp_units_smul, smul_smul, Int.units_mul_self, one_smul,
+        ιMapBifunctorOrZero_eq]
+    · rw [K₁.shape _ _ h₃, Functor.map_zero, zero_app, zero_comp,
+        HomologicalComplex₂.d₁_eq_zero _ _ _ _ _ h₃, zero_comp]
+  · rw [ιMapBifunctor_hom₁_assoc _ _ _ _ _ _ _ _ _ _ rfl]
+    by_cases h₃ : c₁.Rel (c₁.prev i₁) i₁
+    · rw [ιMapBifunctorOrZero_eq _ _ _ _ _ _ _ (by rw [← ComplexShape.prev_π₁ c₂ c h₃, h]),
+        Linear.units_smul_comp, assoc, assoc, HomologicalComplex₂.ι_D₁,
+        HomologicalComplex₂.d₁_eq _ _ h₃ _ _ h, Linear.comp_units_smul,
+        Linear.comp_units_smul, smul_smul, Int.units_mul_self, one_smul]
+      dsimp
+      rw [NatTrans.naturality_assoc]
+      rfl
```

Comment:
Looks like there is some missing material on some subset of `HomologicalComplex₂.ιTotal`, `Functor.mapBifunctorHomologicalComplex`, `GradedObject.ιMapObj`, `GradedObject.mapBifunctor F I₁ I₂` and `ComplexShape.π` here that prevents `simp` from solving this. How hard do you think it would be here to provide a simp-normal form?

Reply:
Thanks very much for this suggestion! It was not so easy to set up, but in the commit https://github.com/leanprover-community/mathlib4/pull/10800/commits/926b86f63c256c17e27a54946974dd17da3e876d I have put more isolation between the API of graded objects and that of the total complex. Now, the proofs in this file `BifunctorHomotopy` are just sequences of `rw`.